proto/substrait/algebra.proto

```diff
@@ -859,6 +857,13 @@ message Expression {
       int32 scale = 3;
     }
 
+    message PrecisionTimestamp {
+      // Sub-second precision, 0 means the value given is in seconds, 3 is milliseconds, 6 microseconds, 9 is nanoseconds
+      int32 precision = 1;
+      // Time passed since 1970-01-01 00:00:00.000000 in UTC for PrecisionTimestampTZ and unspecified timezone for PrecisionTimestamp
```

Comment:
```suggestion
      // Time passed since 1970-01-01 00:00:00.000000 in UTC
```
Sorry, one small clarification.  Timestamps are always in UTC, regardless of whether or not they are `Timestamp` or `TimestampTZ`.

Reply:
Hmm is that right? I think the docs for Timestamp and TimestampTZ disagree, also that’s just the difference between the TZ version and the non-TZ version. The TZ version doesn’t carry any timezone with it, but it counts time since utc epoch so it specifies a definite point in time - while the non-TZ timestamp doesn’t, and is only useful relative to another no -TZ timestamp.

At least that’s the way I understand them, and it aligns with Spark and DataFusion timestamp types.

Reply:
> while the non-TZ timestamp doesn’t, and is only useful relative to another no -TZ timestamp.

My understanding is that you can still do certain things to a non-TZ timestamp:

* Display the wall-clock time represented by the timestamp (e.g. 10PM on Thursday, July 11th, 2024)
* Assign a timezone to it to convert it to a -TZ timestamp

In order for a consumer to do either of those things there needs to be some meaning behind the literal beyond "just some random number that has appropriate relative distance from other values in this column"

In pretty much all cases I'm aware of the non-TZ timstamps are still stored as UTC instants where the wall clock (in a UTC time zone) would display the desired value.

Reply:
> In pretty much all cases I'm aware of the non-TZ timstamps are still stored as UTC instants where the wall clock (in a UTC time zone) would display the desired value.

I checked Spark, DataFusion, Parquet and Postgres docs, and as far as I understand them they all say the non-TZ timestamp bears no notion of timezone at all (ie. it's time passed since _some_ epoch, not UTC epoch), while the yes-TZ timestamp is specifically time passed since UTC epoch. 

> My understanding is that you can still do certain things to a non-TZ timestamp:
> - Display the wall-clock time represented by the timestamp (e.g. 10PM on Thursday, July 11th, 2024)
> - Assign a timezone to it to convert it to a -TZ timestamp

Yes, the non-TZ timestamp does indicate wall clock time - but it will represent the same wall clock time of e.g. 10PM on Thursday, July 11th, 2024 no matter in which timezone you look at it. And you can convert it to a -TZ timestamp by providing information about which timezone's wall clock time it corresponds to.

> In order for a consumer to do either of those things there needs to be some meaning behind the literal beyond "just some random number that has appropriate relative distance from other values in this column"

Well yes, it's not just a random number, but time passed since 1.1.1970.

Reply:
> Well yes, it's not just a random number, but time passed since 1.1.1970.

Ok.  So if a producer encodes the literal "10PM on Thursday, July 11th, 2024" as 1720735200 (seconds since 1.1.1970 in UTC) and my consumer encodes the literal "10PM on Thursday, July 11th, 2024" as 1720760400 (seconds since 1.1.1970 in PT) then I think we are going to have a problem.

Often people say "seconds since the UNIX epoch" but the UNIX epoch is defined as 1.1.1970 UTC and not 1.1.1970 in some arbitrary time zone.

> the non-TZ timestamp bears no notion of timezone at all

Correct.  We are supposed to think of these values semantically as a wall clock time.  However, we still need to agree on how we encode wall clock times into a proto literal.  If we represent wall clock times as ISO-8601 strings (with no TZ part) then it would be fine.  There would be no need to mention time zones.  However, if we choose to encode wall clock times as a u64 then we must mention how that works.  The way that works is by encoding "the wall clock time X" as "seconds since the epoch (UTC) where a wall clock in a UTC time zone would display X"

Reply:
> Ok. So if a producer encodes the literal "10PM on Thursday, July 11th, 2024" as 1720735200 (seconds since 1.1.1970 in UTC) and my consumer encodes the literal "10PM on Thursday, July 11th, 2024" as 1720760400 (seconds since 1.1.1970 in PT) then I think we are going to have a problem.

Yep, if using the non-TZ version of timestamps created with different assumptions, the values are not comparable. That's why the TZ version exists.

> Often people say "seconds since the UNIX epoch" but the UNIX epoch is defined as 1.1.1970 UTC and not 1.1.1970 in some arbitrary time zone.

Yep, that's exactly what the (Precision)TimestampTz is for.

> The way that works is by encoding "the wall clock time X" as "seconds since the epoch (UTC) where a wall clock in a UTC time zone would display X"

That doesn't necessitate specifying UTC. You can as well say "seconds since the epoch (in arbitrary timezone TZ) where a wall clock in the same timezone TZ would display X". That's what I assume is the point of the no-TZ timestamp :)

Reply:
> That doesn't necessitate specifying UTC. You can as well say "seconds since the epoch (in arbitrary timezone TZ) where a wall clock in the same timezone TZ would display X". That's what I assume is the point of the no-TZ timestamp :)

Very well then.  You are a consumer.  I have sent you a plan with a `PrecisionTimestamp` literal.  The proto is:

```
{
  precision: 0,
  value: 1720794104
}
```

Can you please format this as a string of the form `MONTH/DAY/YEAR HOURS:MINUTES:SECONDS`?

Reply:
> Very well then. You are a consumer. I have sent you a plan with a `PrecisionTimestamp` literal. The proto is:
> 
> ```
> {
>   precision: 0,
>   value: 1720794104
> }
> ```
> 
> Can you please format this as a string of the form `MONTH/DAY/YEAR HOURS:MINUTES:SECONDS`?

I'm not sure I understand where you're getting with this, but I believe that'd be `July 12, 2024 14:21:44` (well, with a slightly different format string).

Reply:
I'm afraid that answer is wrong.  The correct answer is `July 12, 2024 7:21:44 AM`.

I believe I followed the encoding rules correctly.  1720794104 is the number of seconds that have passed since 1.1.1970 PT.

Reply:
Actually I think I see my mistake.  I'm thinking we are interpreting this as

"The time shown on a wall clock in UTC after X seconds have passed since Y" (in which case Y's timezone matters)

You are perhaps interpreting it as

"The time shown on a wall clock in timezone Z after X seconds have passed since Y" (in which case Y's time zone doesn't matter as long as Z matches it)

Reply:
I think you're right in how I interpret it 👍